test/stakingPoolDAO_spec.js

```diff
@@ -0,0 +1,274 @@
+      votesN = await StakingPool.methods.votes(proposalId, false).call();
+      
+
+      assert.strictEqual(votesY, "10000000000");
+      assert.strictEqual(votesN, "0");
+
+      receipt = await StakingPool.methods.vote(proposalId, false).send({from: richard});
+
+      votesY = await StakingPool.methods.votes(proposalId, true).call(); 
+      votesN = await StakingPool.methods.votes(proposalId, false).call();
+      
+      assert.strictEqual(votesY, "0");
+      assert.strictEqual(votesN, "10000000000");
+    });
+
+    it("voting is only valid during the period it is active", async () => {
+      const toSend = await StakingPool.methods.vote(proposalId, false);
+
+      await toSend.send({from: richard});
+
+      // Mine 20 blocks
+      for(let i = 0; i < 20; i++){
+        await mineAtTimestamp(12345678);
+      }
+
+      await assert.reverts(toSend, {from: richard}, "Returned error: VM Exception while processing transaction: revert Proposal has already ended");
+    });
+
+    it("check that vote result matches what was voted", async () => {
+      await StakingPool.methods.vote(proposalId, true).send({from: eric});
+      await StakingPool.methods.vote(proposalId, true).send({from: michael});
+      await StakingPool.methods.vote(proposalId, false).send({from: pascal});
+      await StakingPool.methods.vote(proposalId, false).send({from: richard});
+
+      // Mine 20 blocks
+      for(let i = 0; i < 20; i++){
+        await mineAtTimestamp(12345678);
+      }
+
+      let votesY = await StakingPool.methods.votes(proposalId, true).call(); 
+      let votesN = await StakingPool.methods.votes(proposalId, false).call();
+
+      const result = await StakingPool.methods.isProposalApproved(proposalId).call();
+      
+      assert.strictEqual(votesY, "30000000000");
+      assert.strictEqual(votesN, "20000000000");
+      assert.strictEqual(result.approved, true);
+      assert.strictEqual(result.executed, false);
+    });
+  });
+
+  describe("proposal execution", () => {
+    let proposalId;
+    beforeEach(async () => {
+      const receipt = await StakingPool.methods.addProposal("0x00000000000000000000000000000000000000AA", 12345, "0x", "0x").send({from: richard});
+      proposalId = receipt.events.NewProposal.returnValues.proposalId;
+    });
+
+    it("active voting proposals cant be executed", async () => {
+      await StakingPool.methods.vote(proposalId, true).send({from: richard});
+      const toSend = StakingPool.methods.executeTransaction(proposalId);
+      await assert.reverts(toSend, {from: richard}, "Returned error: VM Exception while processing transaction: revert Voting is still active");
+    });
+
+    it("unapproved proposals cant be executed", async () => {
+      await StakingPool.methods.vote(proposalId, false).send({from: richard});
+      
+      // Mine 20 blocks
+      for(let i = 0; i < 20; i++){
+        await mineAtTimestamp(12345678);
```

Comment:
Just curious, wouldn't it be better to use `increaseTime` or another way? Because this slows the test a lot I think

Reply:
It slows it down a little, but nothing too serious. See the time each test took here:
```
 PASS  StakingPoolDAO contract functionality contract should be owned by itself > 0.009s > 4,379,858 gas
 PASS  StakingPoolDAO proposal creation non token holders can not submit proposals > 0.023s > 0 gas
 PASS  StakingPoolDAO proposal creation token holders can create proposals > 0.049s > 157,209 gas
 PASS  StakingPoolDAO voting only those having balance can vote on a proposal > 0.105s > 394,012 gas
 PASS  StakingPoolDAO voting accounts can vote more than once and results are affected accordingly > 0.152s > 247,079 gas
 PASS  StakingPoolDAO voting voting is only valid during the period it is active > 0.192s > 199,166 gas
 PASS  StakingPoolDAO voting check that vote result matches what was voted > 0.305s > 389,514 gas
 PASS  StakingPoolDAO proposal execution active voting proposals cant be executed > 0.089s > 198,952 gas
 PASS  StakingPoolDAO proposal execution unapproved proposals cant be executed > 0.19s > 198,950 gas
 PASS  StakingPoolDAO proposal execution approved proposals can be executed > 0.234s > 282,689 gas
 PASS  StakingPoolDAO proposal execution approved proposals can't be executed twice > 0.234s > 257,689 gas
 PASS  StakingPoolDAO proposal execution approved proposals can't be executed after they expire > 0.326s > 198,952 gas
 PASS  StakingPoolDAO proposal execution proposals can execute contract functions > 0.269s > 532,015 gas
```

I'm currently using block numbers for the contraints on the voting process, but could be changed to timestamps. (I just find using blocknumbers more elegant :) )

Reply:
Ah ok, it's not that bad then 👍